src/runtime_src/core/edge/user/aie/aie.cpp

```diff
@@ -392,55 +477,183 @@ start_profiling(int option, const std::string& port1_name, const std::string& po
-        u8 counterId = acquiredResource.id;
+  XAie_LocType shim_tile1;
+  XAie_StrmPortIntf mode1;
+  uint8_t stream_id1;
+  XAie_LocType shim_tile2;
+  XAie_StrmPortIntf mode2;
+  uint8_t stream_id2;
 
-        XAie_PerfCounterReset(devInst, acquiredResource.loc, XAieModuleType, counterId);
-        XAie_PerfCounterResetControlReset(devInst, acquiredResource.loc, XAieModuleType, counterId);
-
-        if (acquiredResource.module == Resources::pl_module)
-          Resources::AIE::getShimTile(acquiredResource.loc.Col)->plModule.releasePerformanceCounter(phdl, counterId);
-        else if (acquiredResource.module == Resources::core_module)
-          Resources::AIE::getAIETile(acquiredResource.loc.Col, acquiredResource.loc.Row - 1)->coreModule.releasePerformanceCounter(phdl, counterId);
-      } else if (acquiredResource.resource == Resources::stream_switch_event_port) {
-        u8 eventPortId = acquiredResource.id;
+  get_profiling_config(port1_name, shim_tile1, mode1, stream_id1);
+  get_profiling_config(port2_name, shim_tile2, mode2, stream_id2);
 
-        XAie_EventSelectStrmPortReset(devInst, acquiredResource.loc, eventPortId);
+  int handleId = eventRecords.size();
+  int eventPortId1 = Resources::AIE::getShimTile(shim_tile1.Col)->plModule.requestStreamEventPort(handleId);
+  int counterId1 = Resources::AIE::getShimTile(shim_tile1.Col)->plModule.requestPerformanceCounter(handleId);
+  int eventPortId2 = Resources::AIE::getShimTile(shim_tile2.Col)->plModule.requestStreamEventPort(handleId);
+  int counterId2 = Resources::AIE::getShimTile(shim_tile2.Col)->plModule.requestPerformanceCounter(handleId);
+
+  if (counterId1 >= 0 && eventPortId1 >= 0 && counterId2 >= 0 && eventPortId2 >= 0) {
+    if (shim_tile1.Col == shim_tile2.Col) {
+      XAie_EventSelectStrmPort(devInst, shim_tile1, (uint8_t)eventPortId1, mode1, SOUTH, stream_id1);
+      XAie_PerfCounterControlSet(devInst, shim_tile1, XAIE_PL_MOD, (uint8_t)counterId1, XAIE_EVENT_USER_EVENT_0_PL, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId1]);
+      XAie_EventSelectStrmPort(devInst, shim_tile2, (uint8_t)eventPortId2, mode2, SOUTH, stream_id2);
+      XAie_PerfCounterControlSet(devInst, shim_tile2, XAIE_PL_MOD, (uint8_t)counterId2, XAIE_EVENT_USER_EVENT_0_PL, XAIETILE_EVENT_SHIM_PORT_RUNNING[eventPortId2]);
+      XAie_EventGenerate(devInst, shim_tile1, XAIE_PL_MOD, XAIE_EVENT_USER_EVENT_0_PL);
+      eventRecords.push_back({ IO_STREAM_START_DIFFERENCE_CYCLES,
+                  { { shim_tile1, Resources::pl_module, Resources::performance_counter, (size_t)counterId1 },
+                  { shim_tile2, Resources::pl_module, Resources::performance_counter, (size_t)counterId2 },
+                  { shim_tile1, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId1 },
+                  { shim_tile2, Resources::pl_module, Resources::stream_switch_event_port, (size_t)eventPortId2 } } });
+      handle = handleId;
```

Comment:
Is this code duplication from earlier around ~492?
Functions in this file could benefit a lot from refactoring, they are much too long and completely without comments.  Impossible to guess what's going on.